extensions/stackdriver/stackdriver.cc

```diff
@@ -311,34 +343,82 @@ void StackdriverRootContext::record() {
+      outbound, useHostHeaderFallback(), &request_info,
       flatbuffers::GetString(destination_node_info.namespace_()));
   ::Extensions::Stackdriver::Metric::record(
-      isOutbound(), local_node, peer_node, request_info,
-      !config_.disable_http_size_metrics());
+      outbound, local_node, peer_node, request_info,
+      !config_.disable_http_size_metrics(), enableTCPMetrics(),
+      false /* is_tcp */);
   if (enableServerAccessLog() && shouldLogThisRequest()) {
     ::Wasm::Common::populateExtendedHTTPRequestInfo(&request_info);
     logger_->addLogEntry(request_info, peer_node);
   }
   if (enableEdgeReporting()) {
     std::string peer_id;
-    if (!getValue({::Wasm::Common::kDownstreamMetadataIdKey}, &peer_id)) {
-      LOG_DEBUG(absl::StrCat(
-          "cannot get metadata for: ", ::Wasm::Common::kDownstreamMetadataIdKey,
-          "; skipping edge."));
+    if (!getPeerId(peer_id)) {
+      LOG_DEBUG(absl::StrCat("cannot get metadata for: ", peer_id,
+                             "; skipping edge."));
       return;
     }
     edge_reporter_->addEdge(request_info, peer_id, peer_node);
   }
 }
 
+bool StackdriverRootContext::recordTCP(uint32_t id) {
+  const bool outbound = isOutbound();
+  std::string peer_id;
+  getPeerId(peer_id);
+  std::string peer;
+  bool peer_found = getValue(
+      {outbound ? kUpstreamMetadataKey : kDownstreamMetadataKey}, &peer);
+  const ::Wasm::Common::FlatNode& peer_node =
+      *flatbuffers::GetRoot<::Wasm::Common::FlatNode>(
+          peer_found ? peer.data() : empty_node_info_.data());
+  const ::Wasm::Common::FlatNode& local_node = getLocalNode();
+  const ::Wasm::Common::FlatNode& destination_node_info =
+      outbound ? peer_node : local_node;
+
+  if (tcp_request_queue_[id] == nullptr) {
+    return false;
+  }
+  ::Wasm::Common::RequestInfo request_info = *tcp_request_queue_[id];
+  // For TCP, if peer metadata is not available, peer id is set as not found.
+  // Otherwise, we wait for metadata exchange to happen before we report  any
+  // metric.
+  // We keep waiting if response flags is zero, as that implies, there has
+  // been no error in connection.
+  uint64_t response_flags = 0;
+  getValue({"response", "flags"}, &response_flags);
+  if (!peer_found && peer_id != ::Wasm::Common::kMetadataNotFoundValue &&
+      response_flags == 0) {
+    return false;
+  }
+  if (!request_info.is_populated) {
+    ::Wasm::Common::populateTCPRequestInfo(
+        outbound, &request_info,
+        flatbuffers::GetString(destination_node_info.namespace_()));
+  }
+  // Record TCP Metrics.
+  ::Extensions::Stackdriver::Metric::record(
+      outbound, local_node, peer_node, request_info,
+      !config_.disable_http_size_metrics(), enableTCPMetrics(),
+      true /* is_tcp */);
+
+  // Add LogEntry to Logger. Log Entries are batched and sent on timer
+  // to Stackdriver Logging Service.
+  if (enableServerAccessLog() && shouldLogThisRequest()) {
```

Comment:
Just for this PR.. In Access Log PR.. this will be changed to enableTCPServerAccessLog... we currently get access logs for TCP connection at the end and thus didn't want to remove that.

Reply:
If the other PR doesn't get in in time, we will loose the access log we were getting for TCP before...

Reply:
Removed.. yeah this wont happen unless this filter is added in TCP chain...